shell/browser/api/electron_api_utility_process.cc

```diff
@@ -0,0 +1,611 @@
+  if (GetAllUtilityProcessWrappers().Lookup(pid_))
+    GetAllUtilityProcessWrappers().Remove(pid_);
+  node_service_remote_.reset();
+  // Shutdown output readers before exit event..
+  if (stdout_reader_.get())
+    PipeIOBase::Shutdown(std::move(stdout_reader_));
+  if (stderr_reader_.get())
+    PipeIOBase::Shutdown(std::move(stderr_reader_));
+  // Emit 'exit' event
+  Emit("exit", 0 /* exit_code */);
+  Unpin();
+}
+
+void UtilityProcessWrapper::PostMessage(gin::Arguments* args) {
+  if (!node_service_remote_.is_connected())
+    return;
+
+  blink::TransferableMessage transferable_message;
+  v8::Local<v8::Value> message_value;
+  if (args->GetNext(&message_value)) {
+    if (!electron::SerializeV8Value(args->isolate(), message_value,
+                                    &transferable_message)) {
+      // SerializeV8Value sets an exception.
+      return;
+    }
+  }
+
+  v8::Local<v8::Value> transferables;
+  std::vector<gin::Handle<MessagePort>> wrapped_ports;
+  if (args->GetNext(&transferables)) {
+    if (!gin::ConvertFromV8(args->isolate(), transferables, &wrapped_ports)) {
+      args->isolate()->ThrowException(v8::Exception::Error(
+          gin::StringToV8(args->isolate(), "Invalid value for transfer")));
+      return;
+    }
+  }
+
+  bool threw_exception = false;
+  transferable_message.ports = MessagePort::DisentanglePorts(
+      args->isolate(), wrapped_ports, &threw_exception);
+  if (threw_exception)
+    return;
+
+  node_service_remote_->ReceivePostMessage(std::move(transferable_message));
+}
+
+bool UtilityProcessWrapper::Kill() const {
+  if (pid_ == base::kNullProcessId)
+    return 0;
+  base::Process process = base::Process::Open(pid_);
+  bool result = process.Terminate(content::RESULT_CODE_NORMAL_EXIT, false);
+  // Refs https://bugs.chromium.org/p/chromium/issues/detail?id=818244
+  // Currently utility process is not sandboxed which
+  // means Zygote is not used on linux, refs
+  // content::UtilitySandboxedProcessLauncherDelegate::GetZygote.
+  // If sandbox feature is enabled for the utility process, then the
+  // process reap should be signaled through the zygote via
+  // content::ZygoteCommunication::EnsureProcessTerminated.
+  base::EnsureProcessTerminated(process.Duplicate());
+  return result;
+}
+
+v8::Local<v8::Value> UtilityProcessWrapper::GetOSProcessId(
+    v8::Isolate* isolate) const {
+  if (pid_ == base::kNullProcessId)
+    return v8::Undefined(isolate);
+  return gin::ConvertToV8(isolate, pid_);
+}
+
+void UtilityProcessWrapper::ReceivePostMessage(
+    blink::TransferableMessage message) {
+  v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Local<v8::Value> message_value =
+      electron::DeserializeV8Value(isolate, message);
+  Emit("message", message_value);
+}
+
+void UtilityProcessWrapper::HandleMessage(ReaderType type,
+                                          std::vector<uint8_t> message) {
```

Comment:
`HandleMessage` seems like a weird name for this.

Reply:
https://github.com/electron/electron/pull/34980/commits/4a3a57985a314f72bf2afa0b8280d812aeec9c6a